pkg/kudoctl/cmd/diagnostics/collectors.go

```diff
@@ -0,0 +1,132 @@
+package diagnostics
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"reflect"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/kudobuilder/kudo/pkg/kudoctl/clog"
+)
+
+// Ensure collector is implemented
+var _ collector = &resourceCollector{}
+
+// resourceCollector - collector interface implementation for Kubernetes resources (runtime objects)
+type resourceCollector struct {
+	loadResourceFn func() (runtime.Object, error)
+	name           string               // object kind used to describe the error
+	parentDir      stringGetter         // parent dir to attach the printer's output
+	failOnError    bool                 // define whether the collector should return the error
+	callback       func(runtime.Object) // will be called with the retrieved resource after collection to update shared context
+	printMode      printMode
+}
+
+// collect - load a resource and send either the resource or collection error to printer
+// return error if failOnError field is set to true
+// if failOnError is true, finding no object(s) is treated as an error
+func (c *resourceCollector) collect(printer *nonFailingPrinter) error {
+	clog.V(4).Printf("Collect Resource %s in parent dir %s", c.name, c.parentDir())
+	obj, err := c._collect(c.failOnError)
+	if err != nil {
+		printer.printError(err, c.parentDir(), c.name)
+		if c.failOnError {
+			return err
+		}
+	}
+	if obj != nil {
+		printer.printObject(obj, c.parentDir(), c.printMode)
+	}
+	return nil
+}
+
+func emptyResult(obj runtime.Object) bool {
+	return obj == nil || reflect.ValueOf(obj).IsNil() || (meta.IsListType(obj) && meta.LenList(obj) == 0)
+}
+
+func (c *resourceCollector) _collect(failOnError bool) (runtime.Object, error) {
+	obj, err := c.loadResourceFn()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve object(s) of kind %s: %v", c.name, err)
+	}
+	if emptyResult(obj) {
+		obj = nil
+		if failOnError {
+			return nil, fmt.Errorf("no object(s) of kind %s retrieved", c.name)
+		}
+	}
```

Comment:
_return_ statement is probably missing, _if_ will fallthrough (unlike _case_) and callback may be involved on null.